File: go/0131M-分割回文串.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-13 10:42:14
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

func partition(s string) [][]string {
	ans := make([][]string, 0)
	str := make([]string, 0)
	check := func(start, end int) bool {
		for i, j := start, end; i < j; i++ {
			if s[i] != s[j] {
				return false
			}
			j--
		}
		return true
	}
	var dfs func(index int)
	dfs = func(index int) {
		if index >= len(s) {
			tmp := make([]string, len(str))
			copy(tmp, str)
			ans = append(ans, tmp)
			return
		}
		for i := index; i < len(s); i++ {
			if check(index, i) {
				str = append(str, s[index:i+1])
			} else {
				continue
			}
			dfs(i + 1)
			str = str[:len(str)-1]
		}
	}
	dfs(0)
	return ans
}

File: cpp/0131M-分割回文串.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-27 13:46:19
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    vector<vector<string>> ans;
    vector<string> tmp;

    bool check(string s, int start, int end) {
        for (int i = start, j = end; i < j; i++, j--) {
            if (s[i] != s[j]) return false;
        }
        return true;
    }

    void dfs(string s, int idx) {
        if (idx >= s.size()) {
            ans.push_back(tmp);
            return;
        }
        for (int i = idx; i < s.size(); i++) {
            if (!check(s, idx, i)) continue;
            tmp.push_back(s.substr(idx, i - idx + 1));
            dfs(s, i + 1);
            tmp.pop_back();
        }
    }

    vector<vector<string>> partition(string s) {
        dfs(s, 0);
        return ans;
    }
};